### PR TITLE
Issue602 stack overflow

### LINKED ISSUE #602: Bug: StackOverflowException when test calls recursive method in 3.7.
Using .NET Framework 4.8.
This builds fine using NUnit.Analyzers 3.5.0 but fails in 3.7.0 (and 3.8.0):
```csharp
using System;
using System.Collections.Generic;

using NUnit.Framework;

namespace Namespace
{
    internal class Fixture
    {
        private Dictionary<string, bool> _values = new Dictionary<string, bool>();

        [Test]
        public void MinimalRepro()
        {
            Recurse("help");
        }

        private void Recurse(string one, bool keepGoing=true)
        {
            _values[one] = keepGoing;
            if (keepGoing)
            {
                Recurse(one, false);
            }
        }
    }
}
```
Error looks like
image
I don't have time to actually debug the analyser right now but I hope this repro is small enough to make it easy.

### PR DESCRIPTION
**Pull request by manfred-brands**
Fixes #602

### CODE REVIEW

File: src/nunit.analyzers/DisposeFieldsAndPropertiesInTearDown/DisposeFieldsAndPropertiesInTearDownAnalyzer.cs [hunk 1]
```diff
@@ -541,6 +544,22 @@ public Parameters(SemanticModel model, INamedTypeSymbol type, ImmutableHashSet<s
                     SymbolEqualityComparer.Default.Equals(calledMethod.ContainingType, this.type);
             }
 
+            public void ResetMethodCallVisits() => this.visitedMethods.Clear();
+
+            public bool IsLocalNotYetVisitedMethodCall(
+                InvocationExpressionSyntax invocationExpression,
+                [NotNullWhen(true)] out IMethodSymbol? calledMethod)
+            {
+                if (this.IsLocalMethodCall(invocationExpression, out calledMethod) &&
```
Comment (andrewimcclement) [inline src/nunit.analyzers/DisposeFieldsAndPropertiesInTearDown/DisposeFieldsAndPropertiesInTearDownAnalyzer.cs:553]: The whole method simplifies to
```
return this.IsLocalMethodCall(invocationExpression, out calledMethod) &&
       this.visitedMethods.Add(calledMethod);
```

File: src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldOrPropertyIsUninitializedSuppressor.cs [hunk 2]
```diff
@@ -190,7 +195,11 @@ public override void ReportSuppressions(SuppressionAnalysisContext context)
             if (method?.Parent == classDeclaration)
             {
                 // We only get here if the method is in our source code and our class.
-                return IsAssignedIn(model, classDeclaration, method, fieldOrPropertyName);
+                if (!visitedMethods.Contains(method))
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldOrPropertyIsUninitializedSuppressor.cs]: Thanks, excellent suggestion. Code updated.

File: src/nunit.analyzers/DisposeFieldsAndPropertiesInTearDown/DisposeFieldsAndPropertiesInTearDownAnalyzer.cs (hunk 1 above)
Comment (manfred-brands) [reply src/nunit.analyzers/DisposeFieldsAndPropertiesInTearDown/DisposeFieldsAndPropertiesInTearDownAnalyzer.cs]: Thanks, excellent suggestion. Code updated.

**Comment by andrewimcclement** [comment]
@manfred-brands Sorry, I won't be able to review this properly until the weekend.  (I hadn't bothered to have a proper look before since I don't have write access so you would still need another review anyway - at least as far as I understand.)

**Review by mikkelbu — APPROVED** [review]
LGTM

**Pull request merged**